chain/rand/rand.go

```diff
@@ -178,15 +182,27 @@ func (sr *stateRand) GetChainRandomness(ctx context.Context, filecoinEpoch abi.C
 	return sr.getChainRandomness(ctx, filecoinEpoch, true)
 }
 
-func (sr *stateRand) GetBeaconRandomness(ctx context.Context, filecoinEpoch abi.ChainEpoch) ([32]byte, error) {
+func (sr *stateRand) GetBeaconEntry(ctx context.Context, filecoinEpoch abi.ChainEpoch) (*types.BeaconEntry, error) {
 	nv := sr.networkVersionGetter(ctx, filecoinEpoch)
 
-	if nv >= network.Version14 {
-		return sr.getBeaconRandomnessV3(ctx, filecoinEpoch)
-	} else if nv == network.Version13 {
-		return sr.getBeaconRandomnessV2(ctx, filecoinEpoch)
+	if filecoinEpoch > 0 && nv >= network.Version14 {
+		be, err := sr.getBeaconEntryV3(ctx, filecoinEpoch)
+		if err != nil {
+			log.Errorf("failed to get beacon entry as expected: %s", err)
+		}
+		return be, err
+	} else if nv == network.Version13 || filecoinEpoch < 0 {
+		return sr.getBeaconEntryV2(ctx, filecoinEpoch)
+	}
```

Comment:
The `filecoinEpoch` checks change the behaviour here. I'm not sure what the result of `networkVersionGetter(-1)`  is. I'm assuming that it is the network.Version0/1, which would call `getBeaconRandomnessV1`, but it should resolve to genesis randomness either way.

Reply:
They also seem redundant/wierd.
IMO: If we want to special case the `filecoinEpoch < 0`, place it at the top of the function before getting networkVersion and explain in a comment.

Reply:
This tangle is to deal with the odd `<0` check at the top of the current `V3` method, but I think you're right that my replacement logic is iffy.

I've restored the original branch logic to the top-level function and put the `<0` check back into the `V3`. So it should be closer to whatever the original intent of the `<0` check is. I wish I were able to leave a comment about that though because it's very odd.

@Kubuxu would you mind checking out the latest commit here?